Content.Shared/Storage/EntitySystems/BinSystem.cs

```diff
@@ -0,0 +1,173 @@
+
+    public const string BinContainerId = "bin-container";
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<BinComponent, ComponentGetState>(OnGetState);
+        SubscribeLocalEvent<BinComponent, ComponentHandleState>(OnHandleState);
+        SubscribeLocalEvent<BinComponent, ComponentStartup>(OnStartup);
+        SubscribeLocalEvent<BinComponent, MapInitEvent>(OnMapInit);
+        SubscribeLocalEvent<BinComponent, EntRemovedFromContainerMessage>(OnEntRemoved);
+        SubscribeLocalEvent<BinComponent, InteractHandEvent>(OnInteractHand);
+        SubscribeLocalEvent<BinComponent, AfterInteractUsingEvent>(OnAfterInteractUsing);
+    }
+
+    private void OnGetState(EntityUid uid, BinComponent component, ref ComponentGetState args)
+    {
+        args.State = new BinComponentState(component.Items, component.Whitelist, component.MaxItems);
+    }
+
+    private void OnHandleState(EntityUid uid, BinComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not BinComponentState state)
+            return;
+
+        component.Items = new List<EntityUid>(state.Items);
+        component.Whitelist = state.Whitelist;
+        component.MaxItems = state.MaxItems;
+    }
+
+    private void OnStartup(EntityUid uid, BinComponent component, ComponentStartup args)
+    {
+        component.ItemContainer = _container.EnsureContainer<Container>(uid, BinContainerId);
+    }
+
+    private void OnMapInit(EntityUid uid, BinComponent component, MapInitEvent args)
+    {
+        // don't spawn on the client.
+        if (_net.IsClient)
+            return;
+
+        var xform = Transform(uid);
+        foreach (var id in component.InitialContents)
+        {
+            var ent = Spawn(id, xform.Coordinates);
+            if (!TryInsertIntoBin(uid, ent, component))
+            {
+                Logger.Error($"Entity {ToPrettyString(ent)} was unable to be initialized into bin {ToPrettyString(uid)}");
+                return;
+            }
+        }
+    }
+
+    private void OnEntRemoved(EntityUid uid, BinComponent component, EntRemovedFromContainerMessage args)
+    {
+        component.Items.Remove(args.Entity);
+    }
+
+    private void OnInteractHand(EntityUid uid, BinComponent component, InteractHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!_timing.IsFirstTimePredicted)
+            return;
+
+        if (!TryComp<SharedHandsComponent>(args.User, out var hands) || hands.ActiveHand == null)
+            return;
+
+        if (!component.Items.Any())
+            return;
+
+        var toGrab = component.Items.Last();
+        if (!_hands.CanPickupToHand(args.User, toGrab, hands.ActiveHand, false, hands))
+            return;
+
+        if (!TryRemoveFromBin(uid, toGrab, component))
+            return;
+
+        _hands.TryPickup(args.User, toGrab, hands.ActiveHand, handsComp: hands);
```

Comment:
i may be missing something but there are a bunch of repeated checks here imo, moving `TryPickup()` to `TryRemoveFromBin()` should make this a one liner